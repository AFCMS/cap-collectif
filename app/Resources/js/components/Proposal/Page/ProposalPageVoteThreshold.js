// @flow
import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { FormattedMessage } from 'react-intl';
import { ProgressBar } from 'react-bootstrap';
import type { ProposalPageVoteThreshold_step } from './__generated__/ProposalPageVoteThreshold_step.graphql';

type Props = {
  proposal: Object,
  step: ProposalPageVoteThreshold_step,
};

export class ProposalPageVoteThreshold extends React.Component<Props> {

  render() {
    const { proposal, step } = this.props;
    const votesCount = proposal.votesCountByStepId[step.id];
    const votesRemaining = step.voteThreshold - votesCount;
    const votesPercentage = Math.ceil(votesCount * 100 / step.voteThreshold);
    return (
      <div className="proposal__page__vote_threshold">
        <div className="proposal__infos" style={{ marginTop: '-15px' }}>
          <h4>
            {votesPercentage >= 100 ? (
              <FormattedMessage id="proposal.vote.threshold.reached" />
            ) : (
              <FormattedMessage id="proposal.vote.threshold.title" />
            )}
          </h4>
          <p className="proposal__page__vote_threshold__votes">
            <i className="cap cap-hand-like-2-1" />{' '}
            <FormattedMessage
              id="proposal.vote.count"
              values={{
                num: votesCount,
              }}
            />
          </p>
          <ProgressBar
            now={votesPercentage}
            label={`${votesPercentage}%`}
            min={0}
            max={votesPercentage > 100 ? votesPercentage : 100}
            bsStyle="success"
          />
          <div>
            {votesPercentage >= 100 && (
              <FormattedMessage
                id="proposal.vote.threshold.progress_reached"
                values={{
                  num: votesCount,
                  max: step.voteThreshold,
                }}
              />
            )}
            {votesPercentage < 100 && (
              <FormattedMessage
                id="proposal.vote.threshold.progress"
                values={{
                  num: votesRemaining,
                  max: step.voteThreshold,
                }}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
};

export default createFragmentContainer(
  ProposalPageVoteThreshold,
  {
    step: graphql`
      fragment ProposalPageVoteThreshold_step on Step {
        id
        ... on CollectStep {
          voteThreshold
        }
        ... on SelectionStep {
          voteThreshold
        }
      }
    `,
  }
);
