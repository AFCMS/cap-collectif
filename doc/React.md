### How to write localizable components

Just import the appropriate [component](https://github.com/yahoo/react-intl/wiki#the-react-intl-module) from `react-intl`

- For localizable text use
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage).

- For date and time:
[`<FormattedDate>`](https://github.com/yahoo/react-intl/wiki/Components#formatteddate)
[`<FormattedTime>`](https://github.com/yahoo/react-intl/wiki/Components#formattedtime)
[`<FormattedRelative>`](https://github.com/yahoo/react-intl/wiki/Components#formattedrelative)

- For numbers and currencies:
[`<FormattedNumber>`](https://github.com/yahoo/react-intl/wiki/Components#formattednumber)
[`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural)

- If possible, do not use `<FormattedHTMLMessage>`, see how to use *Rich Text Formatting* with
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

- When you need an imperative formatting API, use the [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injectintl) High-Order Component.
