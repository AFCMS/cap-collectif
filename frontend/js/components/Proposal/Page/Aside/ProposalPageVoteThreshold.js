// @flow
import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { FormattedMessage, type IntlShape, injectIntl } from 'react-intl';
import styled, { type StyledComponent } from 'styled-components';
import { ProgressBar } from 'react-bootstrap';
import type { ProposalPageVoteThreshold_step } from '~relay/ProposalPageVoteThreshold_step.graphql';
import type { ProposalPageVoteThreshold_proposal } from '~relay/ProposalPageVoteThreshold_proposal.graphql';
import { isInterpellationContextFromProposal } from '~/utils/interpellationLabelHelper';
import { Card, CategoryCircledIcon } from '~/components/Proposal/Page/ProposalPage.style';
import Icon, { ICON_NAME } from '~/components/Ui/Icons/Icon';
import colors from '~/utils/colors';
import { MetadataRow } from '~/components/Proposal/Page/Aside/ProposalPageMetadata';

type Props = {
  proposal: ProposalPageVoteThreshold_proposal,
  step: ProposalPageVoteThreshold_step,
  showPoints: boolean,
  intl: IntlShape,
};

const ProposalPageVoteThresholdContainer: StyledComponent<{}, {}, HTMLDivElement> = styled.div`
  padding: 20px 30px;

  > h4 {
    margin: 0;
    margin-bottom: 20px;
    font-size: 18px;
    font-weight: 600;
    color: ${colors.darkText};
  }

  > p {
    display: flex;
    justify-content: space-between;
    .vote-counter {
      display: flex;
      align-items: center;
    }
    span {
      margin-left: 10px;
      color: ${colors.secondaryGray};
    }
  }

  /** We use the bootstrap progress-bar, just overriding some style */
  .progress,
  .progress-bar {
    border-radius: 15px !important;
  }
`;

export class ProposalPageVoteThreshold extends React.Component<Props> {
  render() {
    const { step, proposal, showPoints, intl } = this.props;
    // We should use a new query render to fetch votes only from the step
    const votesCount = proposal.votes.totalCount;
    const { voteThreshold } = step;
    let votesRemaining = 0;
    let votesPercentage = 0;
    if (voteThreshold !== null && typeof voteThreshold !== 'undefined' && voteThreshold > 0) {
      votesRemaining = voteThreshold - votesCount;
      votesPercentage = Math.ceil((votesCount * 100) / voteThreshold);
    }
    const isInterpellation = isInterpellationContextFromProposal(proposal);
    return (
      <Card id="ProposalPageVoteThreshold">
        <ProposalPageVoteThresholdContainer>
          <h4>
            {voteThreshold && votesPercentage >= 100 ? (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.vote.threshold.reached'
                    : 'proposal.vote.threshold.reached'
                }
              />
            ) : (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.support.threshold.title'
                    : 'proposal.vote.threshold.title'
                }
              />
            )}
          </h4>
          <p>
            <div className="vote-counter">
              <CategoryCircledIcon size={24} paddingTop={0} paddingLeft={6}>
                <Icon name={ICON_NAME.like} size={14} color={colors.secondaryGray} />
              </CategoryCircledIcon>
              <FormattedMessage
                id={isInterpellation ? 'interpellation.support.count' : 'proposal.vote.count'}
                values={{
                  num: votesCount,
                }}
              />
            </div>
            {proposal && proposal.votes && showPoints && (
              <MetadataRow
                categorySize={24}
                categoryPaddingLeft={5}
                categoryPaddingTop={0}
                name={ICON_NAME.trophy}
                size={14}
                paddingTop={0}
                paddingLeft={3}
                color={colors.secondaryGray}
                ready={!!proposal}
                content={intl.formatMessage(
                  { id: 'count-points' },
                  { num: proposal.votes.totalPointsCount },
                )}
              />
            )}
          </p>

          {voteThreshold && voteThreshold > 0 ? (
            <>
              <ProgressBar
                now={votesPercentage}
                label={`${votesPercentage}%`}
                min={0}
                max={votesPercentage > 100 ? votesPercentage : 100}
                bsStyle="success"
              />
              <div>
                {votesPercentage >= 100 && (
                  <FormattedMessage
                    id={
                      isInterpellation
                        ? 'interpellation.support.threshold.progress_reached'
                        : 'proposal.vote.threshold.progress_reached'
                    }
                    values={{
                      num: votesCount,
                      max: voteThreshold,
                    }}
                  />
                )}
                {votesPercentage < 100 && (
                  <FormattedMessage
                    id={
                      isInterpellation
                        ? 'interpellation.support.threshold.progress'
                        : 'proposal.vote.threshold.progress'
                    }
                    values={{
                      num: votesRemaining,
                      max: voteThreshold,
                    }}
                  />
                )}
              </div>
            </>
          ) : null}
        </ProposalPageVoteThresholdContainer>
      </Card>
    );
  }
}
const container = injectIntl(ProposalPageVoteThreshold);

export default createFragmentContainer(container, {
  proposal: graphql`
    fragment ProposalPageVoteThreshold_proposal on Proposal
      @argumentDefinitions(stepId: { type: "ID!" }) {
      votes(stepId: $stepId, first: 0) {
        totalCount
        totalPointsCount
      }
      ...interpellationLabelHelper_proposal @relay(mask: false)
    }
  `,
  step: graphql`
    fragment ProposalPageVoteThreshold_step on Step {
      ... on CollectStep {
        voteThreshold
      }
      ... on SelectionStep {
        voteThreshold
      }
    }
  `,
});
