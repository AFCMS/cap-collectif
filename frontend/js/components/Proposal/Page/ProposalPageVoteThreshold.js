// @flow
import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { FormattedMessage } from 'react-intl';
import { ProgressBar } from 'react-bootstrap';
import type { ProposalPageVoteThreshold_step } from '~relay/ProposalPageVoteThreshold_step.graphql';
import type { ProposalPageVoteThreshold_proposal } from '~relay/ProposalPageVoteThreshold_proposal.graphql';
import { isInterpellationContextFromProposal } from '~/utils/interpellationLabelHelper';

type Props = {
  proposal: ProposalPageVoteThreshold_proposal,
  step: ProposalPageVoteThreshold_step,
};

export class ProposalPageVoteThreshold extends React.Component<Props> {
  render() {
    const { step, proposal } = this.props;
    // We should use a new query render to fetch votes only from the step
    const votesCount = proposal.votes.totalCount;
    const { voteThreshold } = step;
    if (voteThreshold === null || typeof voteThreshold === 'undefined') {
      return null;
    }
    const votesRemaining = voteThreshold - votesCount;
    const votesPercentage = Math.ceil((votesCount * 100) / voteThreshold);
    const isInterpellation = isInterpellationContextFromProposal(proposal);
    return (
      <div className="proposal__page__vote_threshold">
        <div className="proposal__infos" style={{ marginTop: '-15px' }}>
          <h4>
            {votesPercentage >= 100 ? (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.vote.threshold.reached'
                    : 'proposal.vote.threshold.reached'
                }
              />
            ) : (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.support.threshold.title'
                    : 'proposal.vote.threshold.title'
                }
              />
            )}
          </h4>
          <p className="proposal__page__vote_threshold__votes">
            <i className="cap cap-hand-like-2-1" />{' '}
            <FormattedMessage
              id={isInterpellation ? 'interpellation.support.count' : 'proposal.vote.count'}
              values={{
                num: votesCount,
              }}
            />
          </p>
          <ProgressBar
            now={votesPercentage}
            label={`${votesPercentage}%`}
            min={0}
            max={votesPercentage > 100 ? votesPercentage : 100}
            bsStyle="success"
          />
          <div>
            {votesPercentage >= 100 && (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.support.threshold.progress_reached'
                    : 'proposal.vote.threshold.progress_reached'
                }
                values={{
                  num: votesCount,
                  max: voteThreshold,
                }}
              />
            )}
            {votesPercentage < 100 && (
              <FormattedMessage
                id={
                  isInterpellation
                    ? 'interpellation.support.threshold.progress'
                    : 'proposal.vote.threshold.progress'
                }
                values={{
                  num: votesRemaining,
                  max: voteThreshold,
                }}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default createFragmentContainer(ProposalPageVoteThreshold, {
  proposal: graphql`
    fragment ProposalPageVoteThreshold_proposal on Proposal {
      id
      votes {
        totalCount
      }
      ...interpellationLabelHelper_proposal @relay(mask: false)
    }
  `,
  step: graphql`
    fragment ProposalPageVoteThreshold_step on Step {
      id
      ... on CollectStep {
        voteThreshold
      }
      ... on SelectionStep {
        voteThreshold
      }
    }
  `,
});
