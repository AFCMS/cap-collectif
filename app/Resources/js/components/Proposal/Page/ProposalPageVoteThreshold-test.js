// @flow
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { ProposalPageVoteThreshold } from './ProposalPageVoteThreshold';

describe('<ProposalPageVoteThreshold />', () => {
  const proposal = {
    votes: {
      totalCount: 30,
    },
  };

  const stepWithVoteThreshold = {
    id: '42',
    voteThreshold: 100,
  };

  const stepWithVoteThresholdReached = {
    id: '42',
    voteThreshold: 20,
  };

  it('should render proposal page vote threshold', () => {
    const wrapper = shallow(
      <ProposalPageVoteThreshold proposal={proposal} step={stepWithVoteThreshold} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render proposal page vote threshold reached', () => {
    const wrapper = shallow(
      <ProposalPageVoteThreshold proposal={proposal} step={stepWithVoteThresholdReached} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
