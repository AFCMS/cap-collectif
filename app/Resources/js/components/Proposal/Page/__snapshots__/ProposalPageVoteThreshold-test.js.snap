// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProposalPageVoteThreshold /> should render proposal page vote threshold 1`] = `
<div
  className="proposal__page__vote_threshold"
>
  <div
    className="proposal__infos"
    style={
      Object {
        "marginTop": "-15px",
      }
    }
  >
    <h4>
      <FormattedMessage
        id="proposal.vote.threshold.title"
        values={Object {}}
      />
    </h4>
    <p
      className="proposal__page__vote_threshold__votes"
    >
      <i
        className="cap cap-hand-like-2-1"
      />
       
      <FormattedMessage
        id="proposal.vote.count"
        values={
          Object {
            "num": 0,
          }
        }
      />
    </p>
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="success"
      isChild={false}
      label="0%"
      max={100}
      min={0}
      now={0}
      srOnly={false}
      striped={false}
    />
    <div>
      <FormattedMessage
        id="proposal.vote.threshold.progress"
        values={
          Object {
            "max": 100,
            "num": 100,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`<ProposalPageVoteThreshold /> should render proposal page vote threshold reached 1`] = `
<div
  className="proposal__page__vote_threshold"
>
  <div
    className="proposal__infos"
    style={
      Object {
        "marginTop": "-15px",
      }
    }
  >
    <h4>
      <FormattedMessage
        id="proposal.vote.threshold.title"
        values={Object {}}
      />
    </h4>
    <p
      className="proposal__page__vote_threshold__votes"
    >
      <i
        className="cap cap-hand-like-2-1"
      />
       
      <FormattedMessage
        id="proposal.vote.count"
        values={
          Object {
            "num": 0,
          }
        }
      />
    </p>
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="success"
      isChild={false}
      label="0%"
      max={100}
      min={0}
      now={0}
      srOnly={false}
      striped={false}
    />
    <div>
      <FormattedMessage
        id="proposal.vote.threshold.progress"
        values={
          Object {
            "max": 20,
            "num": 20,
          }
        }
      />
    </div>
  </div>
</div>
`;
